Don't write private key to disk

bovine3dom opened the pull request:
Closes #42.

~~I haven't left writing the private key to disk as an option. The only valid usage I can think of is if you needed to distribute signing across a server farm, which you don't.~~

~~I've cheekily made the public key follow the `name_version.bikey` format since there will be a different one for each version now.~~

~~I've just spotted #47. I'm happy to rebase this on that once it has been merged.~~


---

- Don't write private key to disk unless `reuse_private_key` is set to true in the project configuration
    - If it is set to true, revert to the old behaviour
- Change the default key name to include the version if `reuse_private_key` is false, stick with old default otherwise

- Split copy / sign up into separate functions and do a bit of tidying up (I went a bit off-piste, sorry...)

synixebrett commented:
Looks good, I should have the build refactor done by the end of this next week so I am going to wait until after that to merge this.

jonpas commented on src/build/release.rs:
```diff
@@ -22,26 +22,24 @@ pub fn release(p: &crate::project::Project, version: &String) -> Result<(), Erro
-    }
 
-    // Sign
+    let keyname = format!("{}.{}",p.get_keyname(),version);
```
```suggestion
    // KeyGen
    let keyname = format!("{}.{}",p.get_keyname(),version);
```

jonpas commented on src/build/sign.rs:
```diff
@@ -1,30 +1,34 @@
+    let signame = p.get_signame(&pbo_path);
+
+    let pbo = armake2::pbo::PBO::read(&mut 
+                                      std::fs::File::open(PathBuf::from(
```
What the hell is this indentation! Use variables. :laughing:

bovine3dom replied:
#38 :(

I'm not used to thinking about indentation. I think Vim probably did it for me. I don't really want to make loads of variables when all it's really doing is updating the PBO path. Rust makes it really verbose to do that 🤷‍♂️

jonpas replied:
It's a lot easier to change variables and move them around than 3x nested function call.

bovine3dom commented:
Rebased on current master, should be ready for merge. 

I was getting some very strange errors when I ran it but I think that was because I was changing git branches while using `entr` to rebuild the files as I changed them... I haven't been able to reproduce the errors since, so I don't think they were real. I've since run it quite a few times and used `armake2 verify` to check a few. All seems fine.

jonpas commented on src/build/sign.rs:
```diff
@@ -1,30 +1,48 @@
-    fs::copy(&cpath, format!("releases/{}/@{}/{}/{}", version, modname, folder, pbo))?;
+    fs::copy(
+        &cpath,
+        format!("releases/{}/@{}/{}/{}", version, modname, folder, pbo_path),
```
:thinking:

bovine3dom replied:
What's the `thunk` for? `rustfmt` decided on the newline, not me :(

jonpas replied:
It's hideous. :D

bovine3dom commented:
I'll add a `reuse_private_key` option to the TOML and maybe a flag.

bovine3dom commented on src/build/sign.rs:
```diff
@@ -11,36 +11,39 @@ pub fn copy_sign(
+    let path_posix = path.clone().to_str().unwrap().replace(r#"\"#, "/");
+    let pbo_filename = path_posix.replace((folder.clone() + "/").as_str(), "");
+
+    if !path.ends_with(".pbo") && !pbo_filename.starts_with(p.prefix.as_str()) {
```
This is where I integrated #51. I meant to separate it into a separate commit but apparently I forgot.

bovine3dom commented:
I think this is ready to be reviewed.

Is it worth documenting the weird behaviour you'll get if you specify a dynamic `keyname` but ask us to reuse the key?

synixebrett commented:
Yeah, we should probably use a static name then for the private key. The only reason it would be on disk is if they wanted to reuse it. In that case they would want a static name.

synixebrett commented on docs/project.md:
```diff
@@ -167,6 +167,8 @@ HEMTT will use the specified mod name (without `@`) to form `@mod` folder. Suppo
 
 HEMTT will use the specified key name for `.bikey` and `.biprivatekey` names. Supports [templating](/templating.md).
 
+The default is `{{prefix}}_{{version}}`, unless `reuse_private_key` is `true`, in which case the default is set to `{{prefix}}`
```
It would be nice if this had a table instead, showing the difference between `reuse_private_key` being true or false. This would help clearly show what the default is, and what the output will be in a clear and readable way.

synixebrett commented on docs/project.md:
```diff
@@ -199,3 +202,16 @@ HEMTT will use the specified signature name as part of the full signature (`.bis
+
+If set to `true`, HEMTT will use (and re-use) `releases/keys/{keyname}.biprivatekey`. It will be generated if it doesn't exist.
+
+The default behaviour is to generate a new private key each time and discard it immediately; HEMTT strongly recommends that you only re-use the key if you are making a client-side mod where it will not matter if clients are running different versions of the mod.
```
docisfy has a nice helper for highlighting important info. https://docsify.js.org/#/helpers?id=important-content

It should be used here to highlight the recommendation.

synixebrett commented on src/build/release.rs:
```diff
@@ -18,45 +22,87 @@ pub fn release(p: &crate::project::Project, version: &String) -> Result<(), Erro
+        }
+
+        // Read the private key from disk
+        key = BIPrivateKey::read(
```
I am not a fan of this formatting at all, is this what rustfmt did?

bovine3dom replied:
Yeah, I wouldn't have done this. It's before I increased the line length in the other PR so it'll probably disappear if I run it again.

synixebrett commented on src/build/release.rs:
```diff
@@ -18,45 +22,87 @@ pub fn release(p: &crate::project::Project, version: &String) -> Result<(), Erro
-        if !Path::new(&format!("releases/{}/@{}/{}", version, modname, folder)).exists() {
-            fs::create_dir_all(format!("releases/{}/@{}/{}", version, modname, folder))?;
+        let addonsfolder = format!(
+            "releases/{ver}/@{mod}/{addons}", ver=version, mod=modname, addons=folder
```
The use of variables here seems unnecessary, especially since we don't use them anywhere else. `mod` is also a keyword and shouldn't be used as a variable.

bovine3dom replied:
I don't think these are variables - they just annotate the format! macro. I personally hate the empty {} ones. Happy to remove if you want.

bovine3dom replied:
https://doc.rust-lang.org/std/fmt/#named-parameters

synixebrett replied:
I know they aren't variables outside the format, I just don't see a reason for their use in a simple string where each variable is in order and used once. Especially when they are using different names.

If rust ever adds string interpolation we would definitely use that, but for now, I think named parameters are not needed on simple strings like this.

bovine3dom replied:
It's because the format!/printf macros are alien to me. I find named parameters much more readable. I honestly found the line much easier to understand with it in; I only understood what it did after rewriting it.

If you feel as strongly but in the other direction I'll remove it.

synixebrett replied:
#79 needs a little bit more testing, but I think that'll be a good compromise. There is no redeclaration, or like in your PR, declaring with different names. I am not a fan of changing the name for formatting, but I also think `name = name` looks ridiculous. This is the best we can do with hygienic macros unless rust one days adds string interpolation.

synixebrett commented on src/build/release.rs:
```diff
@@ -18,45 +22,87 @@ pub fn release(p: &crate::project::Project, version: &String) -> Result<(), Erro
         opts.par_iter().for_each(|entry| {
             // TODO split copy and sign
-            if crate::build::sign::copy_sign(&folder, &entry, &p, &version).unwrap() {
+            // for copying, we need to know source path, addons folder and pbo_filename (we could
```
move the start of ( to the next line

synixebrett commented on src/build/sign.rs:
```diff
@@ -1,30 +1,59 @@
+    Ok(true)
+}
+
+pub fn sign(
```
I know this is rustfmt's doing, and the rusty way.

Personally, I am not a fan, but I would be open to switching the entire project to it. That is out of scope here though.

jonpas replied:
Also not a fan.

synixebrett commented on src/build/sign.rs:
```diff
@@ -1,30 +1,59 @@
-    let signame = p.get_signame(&pbo);
-    let keyname = p.get_keyname();
+    let addonsfolder = format!(
+        "releases/{ver}/@{mod}/{addons}", ver=p.version.clone().unwrap(), mod=modname, addons=folder
```
Same as previous variables comment

synixebrett commented on src/project.rs:
```diff
@@ -63,6 +63,8 @@ pub struct Project {
     #[serde(skip_deserializing,skip_serializing)]
     pub template_data: BTreeMap<&'static str, String>,
+
+    pub reuse_private_key: Option<bool>,
```
This should be above the internal variable `template_data`

synixebrett commented on src/project.rs:
```diff
@@ -63,6 +63,8 @@ pub struct Project {
     #[serde(skip_deserializing,skip_serializing)]
     pub template_data: BTreeMap<&'static str, String>,
+
+    pub reuse_private_key: Option<bool>,
```
This should use a serde default flag instead of an `Option` since it will always have a value.

bovine3dom replied:
I struggled with this a bit and got some horrible compiler errors. Could you point me in the right direction?

synixebrett replied:
Unfortunately, serde default needs to be a function.

You can use something like [this](https://github.com/Dynulo/Gluon/blob/master/src/files/config.rs#L48), then we can add [this](https://github.com/Dynulo/Gluon/blob/master/src/files/config.rs#L54) to main.rs and use `crate::dft_false` as our default value. It isn't ideal, but it is better than an option.

synixebrett commented on src/project.rs:
```diff
@@ -93,7 +95,11 @@ impl Project {
+            if self.reuse_private_key.unwrap_or(false) {
+                self.prefix.clone()
+            } else {
+                format!("{}_{}", &self.prefix, &self.version.clone().unwrap())
```
`unwrap` is added a few times in this PR, make sure `unwrap_or_print` is used wherever allowed. That is what prints the nice readable error messages rather.

bovine3dom commented:
Rebased, tidied up and all changes made. Tested and it still works (which almost never happens after I rebase).

synixebrett commented on src/build/release.rs:
```diff
@@ -26,50 +29,83 @@ pub fn release(p: &crate::project::Project, version: &String) -> Result<(), Erro
         fs::create_dir("releases/keys")?;
     }
+
+    let mut key;
```
I'm surprised the compiler allowed this, variables should always be given a value when they are defined. This should just be down on line 54.

bovine3dom replied:
It's just a lazy way of allowing the two cases of `key` to store their result in a variable that the rest of the programme can read. If I put it down further, it'd go out of scope.

I guess the "right" way would be to do `let key = give_me_the_key(reuse_private_key: bool, ...)` and put the if below into it?

I don't really see the issue with how it is currently.

synixebrett replied:
In rust `if` statements can return values. The if should return a value rather than using an empty value.

```rust
let x: int;
if something {
	case1();
	x = 1;
} else {
	case2();
	x = 2;
}
```
should be
```rust
let x = if something {
	case1();
	1
} else {
	case2();
	2
};
```

bovine3dom replied:
Oh, that's magic. Thanks.
